internal/auth/internal/types/traffic_permissions.go

```diff
@@ -31,114 +32,206 @@ func RegisterTrafficPermissions(r resource.Registry) {
+					Name:    "source",
+					Wrapped: errSourceWildcards,
+				}))
 			}
 		}
-		if len(permission.DestinationRules) > 0 {
-			for d, dest := range permission.DestinationRules {
-				wrapDestRuleErr := func(err error) error {
-					return wrapPermissionErr(resource.ErrInvalidListElement{
-						Name:    "destination_rules",
-						Index:   d,
+	}
+	for d, dest := range p.DestinationRules {
+		wrapDestRuleErr := func(err error) error {
+			return wrapErr(resource.ErrInvalidListElement{
+				Name:    "destination_rules",
+				Index:   d,
+				Wrapped: err,
+			})
+		}
+		if (len(dest.PathExact) > 0 && len(dest.PathPrefix) > 0) ||
+			(len(dest.PathRegex) > 0 && len(dest.PathExact) > 0) ||
+			(len(dest.PathRegex) > 0 && len(dest.PathPrefix) > 0) {
+			merr = multierror.Append(merr, wrapDestRuleErr(resource.ErrInvalidListElement{
+				Name:    "destination_rule",
+				Wrapped: errInvalidPrefixValues,
+			}))
+		}
+		if len(dest.Exclude) > 0 {
+			for e, excl := range dest.Exclude {
+				wrapExclPermRuleErr := func(err error) error {
+					return wrapDestRuleErr(resource.ErrInvalidListElement{
+						Name:    "exclude_permission_rules",
+						Index:   e,
 						Wrapped: err,
 					})
 				}
-				if (len(dest.PathExact) > 0 && len(dest.PathPrefix) > 0) ||
-					(len(dest.PathRegex) > 0 && len(dest.PathExact) > 0) ||
-					(len(dest.PathRegex) > 0 && len(dest.PathPrefix) > 0) {
-					err = multierror.Append(err, wrapDestRuleErr(resource.ErrInvalidListElement{
-						Name:    "destination_rule",
+				if (len(excl.PathExact) > 0 && len(excl.PathPrefix) > 0) ||
+					(len(excl.PathRegex) > 0 && len(excl.PathExact) > 0) ||
+					(len(excl.PathRegex) > 0 && len(excl.PathPrefix) > 0) {
+					merr = multierror.Append(merr, wrapExclPermRuleErr(resource.ErrInvalidListElement{
+						Name:    "exclude_permission_rule",
 						Wrapped: errInvalidPrefixValues,
 					}))
 				}
-				if len(dest.Exclude) > 0 {
-					for e, excl := range dest.Exclude {
-						wrapExclPermRuleErr := func(err error) error {
-							return wrapPermissionErr(resource.ErrInvalidListElement{
-								Name:    "exclude_permission_rules",
-								Index:   e,
-								Wrapped: err,
-							})
-						}
-						if (len(excl.PathExact) > 0 && len(excl.PathPrefix) > 0) ||
-							(len(excl.PathRegex) > 0 && len(excl.PathExact) > 0) ||
-							(len(excl.PathRegex) > 0 && len(excl.PathPrefix) > 0) {
-							err = multierror.Append(err, wrapExclPermRuleErr(resource.ErrInvalidListElement{
-								Name:    "exclude_permission_rule",
-								Wrapped: errInvalidPrefixValues,
-							}))
-						}
-					}
-				}
 			}
 		}
 	}
 
-	return err
+	return merr
+}
+
+func sourceHasIncompatibleTenancies(src pbauth.SourceToSpiffe) bool {
+	peerSet := src.GetPeer() != resource.DefaultPeerName
+	apSet := src.GetPartition() != resource.DefaultPartitionName
+	sgSet := src.GetSamenessGroup() != ""
+
+	return (apSet && peerSet) || (apSet && sgSet) || (peerSet && sgSet)
```

Comment:
Because this initially confused me:

- You can't set partition + peer, despite peer being a partitioned concept, because this rule is defined in the context of the destination itself, which exists in a partition and is talking about its *own* peers, so the partition is effectively defaulted.
- sameness groups can use similar logic
- sameness groups contain peers

Really this is treating the 3 concepts like a protobuf `oneof`.